dbal driver

JarekW opened the pull request.

JarekW commented:
@fsi-open what do you think?

chives commented on lib/Driver/Doctrine/DBAL/Extension/Core/EventSubscriber/ResultIndexer.php:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Class contains method called at BindParameters events.
+ */
+class ResultIndexer implements EventSubscriberInterface
```
is this class even used?

norzechowicz commented:
DBAL driver is probably the only driver that datasource needs. 
We created one for internal usage in project I'm working on currently, this is simple usage example:

```php
        $queryBuilder = $this->getObjectManager()->getConnection()->createQueryBuilder();
        $queryBuilder
            ->select('u.id, u.name')
            ->from('users', 'u')
            ->orderBy('u.registration_date', 'DESC');

        $datasource = $factory->createDataSource('doctrine-dbal', [
            'queryBuilder' => $queryBuilder,
            'countField' => 'u.id',
            'indexField' => 'id'
        ], 'users');
```

It can be used with datagrid

```php
        $datagrid->addColumn('username', 'text', [
            'label' => 'Username',
            'field_mapping' => ['[name]']
        ]);
```

however you need to remember to access fields using `[notation]` array access notation. 
I will try to extract our code as soon as I find some time so you can take a look at it and maybe implement some parts of it in your solution.

JarekW commented:
one thing I see in your solution and should implement is the `indexField` that probably is used for key in results. However `countField` possibly could be skipped - look at `Driver/Doctrine/DBAL/Paginator`

norzechowicz commented:
@JarekW have you tested that with more complicated queries that include joins, sub-selects etc?

JarekW commented:
Paginator just wraps whole query and count rows. I don't see (at least for now) why it could not work.
There was tests for SQL query construction (as in ORM part) for joins and aggregates, so I've added there assertions to check result count.

While adding those test some problem came up with how sqlite works in comparison with mysql :/ but better sooner rather than later.

I've added `integer` dbal type to `number` type field because sqlite did't work (mysql did). This change showed that I also need to bind parameter _types_ from original query builder inside Paginator. Unfortunatelly  DBAL <=2.4 doesn't have `getParameterTypes` method on `QueryBuilder`, although it has this property since qb was introduces. There is ugly reflection solution for now (can we require DBAL >=2.5 ?) Any advice?

szymach commented on lib/Driver/Doctrine/DBAL/DBALAbstractField.php:
```diff
@@ -0,0 +1,141 @@
+            })
+            ->setNormalizer('clause', function($options, $value) {
+                return strtolower($value);
+            }
```
Merge this and the next line, remove duplicated ampersand

szymach commented on lib/Driver/Doctrine/DBAL/DBALAbstractField.php:
```diff
@@ -0,0 +1,141 @@
+            $comparison = 'like';
+        }
+
+        $qb->$func($expr->$comparison($fieldName, ":$name"));
```
Wouldn't hurt to check if the `$func` is actually callable on the query builder.

szymach commented on lib/Driver/Doctrine/DBAL/DBALDriver.php:
```diff
@@ -0,0 +1,159 @@
+class DBALDriver extends DriverAbstract
+{
+    /**
+     * Default alias for entity during building query when no alias is specified.
```
The constant name is pretty self-explanatory to me :)

szymach commented on lib/Driver/Doctrine/DBAL/DBALDriver.php:
```diff
@@ -0,0 +1,159 @@
+     *
+     * @var QueryBuilder
+     */
+    private $query;
```
I'd name this `initialQuery` or something

szymach commented on lib/Driver/Doctrine/DBAL/DBALDriver.php:
```diff
@@ -0,0 +1,159 @@
+    private $query;
+
+    /**
+     * Query builder available during preGetResult event.
```
Duplicated description from above.

szymach commented on lib/Driver/Doctrine/DBAL/DBALDriver.php:
```diff
@@ -0,0 +1,159 @@
+    /**
+     * Returns query builder.
+     *
+     * If query is set to null (so when getResult method is NOT executed at the moment) exception is throwed.
```
`is thrown`

szymach commented on lib/Driver/Doctrine/DBAL/DBALFactory.php:
```diff
@@ -0,0 +1,92 @@
+    private $registry;
+
+    /**
+     * Array of extensions.
```
Unnecessary, name and type are explanatory enough.

szymach commented on lib/Driver/Doctrine/DBAL/DBALDriver.php:
```diff
@@ -0,0 +1,159 @@
+            $this->alias = self::DEFAULT_ENTITY_ALIAS;
+        }
+
+        if ($table instanceof QueryBuilder) {
```
Abomination

JarekW replied:
similar in [ORM driver](https://github.com/fsi-open/datasource/blob/master/lib/Driver/Doctrine/ORM/DoctrineDriver.php#L82), 2 parameters?

szymach commented on lib/Driver/Doctrine/DBAL/DBALFactory.php:
```diff
@@ -0,0 +1,92 @@
+        ));
+
+        $this->optionsResolver->setAllowedTypes('qb', array('\Doctrine\DBAL\Query\QueryBuilder', 'null'));
+        $this->optionsResolver->setAllowedTypes('table', array('string', 'null'));
```
Should also allow `'\Doctrine\DBAL\Query\QueryBuilder'`, looking at the constructor method

JarekW replied:
separate option, later chosen [as in orm](https://github.com/fsi-open/datasource/blob/master/lib/Driver/Doctrine/ORM/DoctrineFactory.php#L72)

szymach commented on lib/Driver/Doctrine/DBAL/DBALResult.php:
```diff
@@ -0,0 +1,69 @@
+    public function __construct(Paginator $paginator, $indexField)
+    {
+        if (!is_string($indexField) && !$indexField instanceof \Closure) {
+            throw new \InvalidArgumentException('indexField should be string or \Closure');
```
Add `get_class/gettype` value result accordingly to variable type message.

szymach commented on lib/Driver/Doctrine/DBAL/Extension/Core/Field/Date.php:
```diff
@@ -0,0 +1,25 @@
+ */
+class Date extends DBALAbstractField
+{
+    protected $comparisons = array('eq', 'neq', 'lt', 'lte', 'gt', 'gte', 'in', 'notIn', 'between', 'isNull');
```
Are `eq`, `neq`, `in` and `notIn` comparisons even necessary?

JarekW replied:
same as [orm date](https://github.com/fsi-open/datasource/blob/master/lib/Driver/Doctrine/ORM/Extension/Core/Field/Date.php)

szymach commented on lib/Driver/Doctrine/DBAL/Extension/Core/Field/Number.php:
```diff
@@ -0,0 +1,36 @@
+    public function getDBALType()
+    {
+        /*
+         * when we omit type, doctrine will bind value to statement as \PDO::PARAM_STR.
```
`If the type is ommited, Doctrine will bind the value as \PDO::PARAM_STR. This will result in incorrect results for some engines (like SQLite).`

szymach commented on lib/Driver/Doctrine/DBAL/Paginator.php:
```diff
@@ -0,0 +1,59 @@
+        }
+
+        //dbal <=2.4
+        $class = new \ReflectionClass('Doctrine\DBAL\Query\QueryBuilder');
```
I don't really see another way to do this, so let it be.

szymach commented on lib/Extension/Core/Ordering/Driver/DBALExtension.php:
```diff
@@ -0,0 +1,82 @@
+            throw new \InvalidArgumentException("Field must be an instance of DoctrineField");
+        }
+
+        if ($field->hasOption('field')) {
```
Would've been better using ternarny operator
